ext/dom/element.c

```diff
@@ -1415,76 +1530,65 @@ PHP_METHOD(DOMElement, replaceChildren)
+#define INSERT_ADJACENT_RES_PRE_INSERT_FAILED ((void*) -2)
 
 static xmlNodePtr dom_insert_adjacent(const zend_string *where, xmlNodePtr thisp, dom_object *this_intern, xmlNodePtr otherp)
 {
 	if (zend_string_equals_literal_ci(where, "beforebegin")) {
 		if (thisp->parent == NULL) {
 			return NULL;
 		}
-		if (dom_hierarchy(thisp->parent, otherp) == FAILURE) {
-			php_dom_throw_error(HIERARCHY_REQUEST_ERR, dom_get_strict_error(this_intern->document));
-			return INSERT_ADJACENT_RES_FAILED;
-		}
 		if (!php_dom_adopt_node(otherp, this_intern, thisp->doc)) {
-			return INSERT_ADJACENT_RES_FAILED;
+			return INSERT_ADJACENT_RES_ADOPT_FAILED;
 		}
-		otherp = xmlAddPrevSibling(thisp, otherp);
-	} else if (zend_string_equals_literal_ci(where, "afterbegin")) {
-		if (dom_hierarchy(thisp, otherp) == FAILURE) {
-			php_dom_throw_error(HIERARCHY_REQUEST_ERR, dom_get_strict_error(this_intern->document));
-			return INSERT_ADJACENT_RES_FAILED;
+		if (!php_dom_pre_insert(this_intern->document, otherp, thisp->parent, thisp)) {
+			return INSERT_ADJACENT_RES_PRE_INSERT_FAILED;
 		}
+	} else if (zend_string_equals_literal_ci(where, "afterbegin")) {
 		if (!php_dom_adopt_node(otherp, this_intern, thisp->doc)) {
-			return INSERT_ADJACENT_RES_FAILED;
+			return INSERT_ADJACENT_RES_ADOPT_FAILED;
 		}
-		if (thisp->children == NULL) {
-			otherp = xmlAddChild(thisp, otherp);
-		} else {
-			otherp = xmlAddPrevSibling(thisp->children, otherp);
+		if (!php_dom_pre_insert(this_intern->document, otherp, thisp, thisp->children)) {
+			return INSERT_ADJACENT_RES_PRE_INSERT_FAILED;
 		}
 	} else if (zend_string_equals_literal_ci(where, "beforeend")) {
-		if (dom_hierarchy(thisp, otherp) == FAILURE) {
-			php_dom_throw_error(HIERARCHY_REQUEST_ERR, dom_get_strict_error(this_intern->document));
-			return INSERT_ADJACENT_RES_FAILED;
-		}
 		if (!php_dom_adopt_node(otherp, this_intern, thisp->doc)) {
-			return INSERT_ADJACENT_RES_FAILED;
+			return INSERT_ADJACENT_RES_ADOPT_FAILED;
+		}
+		if (!php_dom_pre_insert(this_intern->document, otherp, thisp, NULL)) {
+			return INSERT_ADJACENT_RES_PRE_INSERT_FAILED;
 		}
-		otherp = xmlAddChild(thisp, otherp);
 	} else if (zend_string_equals_literal_ci(where, "afterend")) {
 		if (thisp->parent == NULL) {
 			return NULL;
 		}
-		if (dom_hierarchy(thisp->parent, otherp) == FAILURE) {
-			php_dom_throw_error(HIERARCHY_REQUEST_ERR, dom_get_strict_error(this_intern->document));
-			return INSERT_ADJACENT_RES_FAILED;
-		}
 		if (!php_dom_adopt_node(otherp, this_intern, thisp->doc)) {
-			return INSERT_ADJACENT_RES_FAILED;
+			return INSERT_ADJACENT_RES_ADOPT_FAILED;
+		}
+		if (!php_dom_pre_insert(this_intern->document, otherp, thisp->parent, thisp->next))  {
+			return INSERT_ADJACENT_RES_PRE_INSERT_FAILED;
 		}
-		otherp = xmlAddNextSibling(thisp, otherp);
 	} else {
 		php_dom_throw_error(SYNTAX_ERR, dom_get_strict_error(this_intern->document));
-		return INSERT_ADJACENT_RES_FAILED;
+		return INSERT_ADJACENT_RES_SYNTAX_FAILED;
 	}
```

Comment:
Should this be an assertion? As it seems this depends on `where` being an invalid value.

Reply:
Why should this be an assertion? `where` is a user-provided string. So you _can_ hit this.